envi/tests/test_arch_arm.py

```diff
@@ -1468,16 +1490,213 @@ class ArmInstructionSet(unittest.TestCase):
     armTestVersion = REV_ARMv7A
     armTestOnce = True
 
+    def test_PSRs(self):
+
+        am = arm.ArmModule()
+        emu = am.getEmulator()
+
+        # there can be only one CPSR
+        emu.setCPSR(47147)
+        self.assertEqual(emu.getCPSR(), 47147)
+        self.assertEqual(emu.getRegister(REG_CPSR), 47147)
+
+        for mode in range(MODE_COUNT):
+            modeval = mode | 0b10000
+            if not modeval in proc_modes:
+                continue
+
+            #print("Testing Mode %x (%r)" % (modeval, proc_modes.get(modeval)[0]))
+            #print("  SPSR")
+            emu.setSPSR(mode, 31337 + mode)
+            self.assertEqual(emu.getSPSR(mode), 31337 + mode)
+
+            regidx = reg_mode_base(modeval) + REG_CPSR
+            ridx = eaar._getRegIdx(regidx, 0)
+            #print("  CPSR(%x): 0x%x -> 0x%x" % (modeval, regidx, ridx))
+            #self.assertEqual(emu._rctx_vals[regidx], 47147)
+            self.assertEqual(emu._rctx_vals[ridx], 47147)
+            self.assertEqual(emu.getRegister(REG_CPSR, mode=modeval), 47147)
+
+        '''
```

Comment:
oh, it's a weird github ui thing.
i see what you're referring to now.

this is the humon test that the unittests are exemplifying.  because this is a complicated lookup table, i'm trying to show ground truth for future us'es to troubleshoot.

Reply:
i am not entirely sure what you're saying, so please explain further.

as for this, this information is specifically pertaining to this unittest to help if it ever fails.

but if you can tell me what file to move this data into, i will gladly comply.

Reply:
ARM registers (banked) are kinda complex, and if the numbering/offset/mapping ever gets off, i want the info to be available for what it was when i merged this in.  i'm not entirely sure where to put that, but here seemed a good place to me at the time.

Reply:
what would you like me to do with this?

Reply:
So, I'm just not entirely sure what exactly is being preserved here. If you're trying to give details about the ARM register structure, I think we could be a lot more clear here, or even just encode that in a more specific test.

But at least personally, ipython output as clarifying documentation isn't particularly helpful. In a perfect world, we could have a section in our docs talking about ARM and it's various structures so that we could refer back to our own (hopefully much clearer than ARM's) writings. But that's also a lot of work and I'm not going to levy that as a thing I need to see. That's mostly just wishful thinking.

So maybe just some words to more completely explain the SPSR structure and what we're doing, or just encode that into the test?

Reply:
gee, i wish we had someplace to put such things!

<facepalm>

moved to `vivisect/notes/envi/arm/registers.md`